plugins/optimization/src/main/java/org/qbicc/plugin/opt/FenceOptimizer.java

```diff
@@ -0,0 +1,463 @@
+        } else if (prev instanceof Store) {
+            Store prev_store = (Store) prev;
+            //
+            // prev_store
+            //  seq_cst
+            //    +
+            //  acquire   <- not needed
+            //   load
+            //
+            if (prev_store.getMode() == MemoryAtomicityMode.SEQUENTIALLY_CONSISTENT
+                    && load.getMode() == MemoryAtomicityMode.ACQUIRE) {
+                load.setMode(MemoryAtomicityMode.UNORDERED);
+            }
+        }
+
+        return true;
+    }
+
+    private boolean weaken(Fence fence, Node prev) {
+        if (prev instanceof Store) {
+            Store prev_store = (Store) prev;
+            //
+            // prev_store
+            // seq_cst
+            //    +
+            // release <- not needed
+            //
+            if (prev_store.getMode() == MemoryAtomicityMode.SEQUENTIALLY_CONSISTENT
+                    && fence.getAtomicityMode() == MemoryAtomicityMode.RELEASE) {
+                fence.setAtomicityMode(MemoryAtomicityMode.UNORDERED);
+            }
+        }
+
+        return true;
+    }
+
+    /**
+     * Try to weaken fence by calculating the weakest fence in the incoming set.
+     *
+     * @return {@code true} if optimization should finish because of either successful optimization or
+     *         giving up optimization due to some reasons, {@code false} otherwise .
+     *
+     */
+    private boolean weaken(Node n, Set<Node> incomingSet) {
+        if (incomingSet == null) {
+            return true;
+        }
+
+        Set<Node> functionCallFreeSet = new HashSet<Node>();
+        boolean isEmptySet = false;
+        for (Node prev : incomingSet) {
+            Object[] ret = new Object[1];
+            if (getIncomingIfFunctionCall(prev, ret)) {
+                Set<Node> set = (Set<Node>) ret[0];
+                if (set == null) { // Couldn't get the information. Give up optimizing node n.
+                    return true;
+                }
+                if (set.size() == 0) { // No fence exists.
+                    isEmptySet = true;
+                    continue;
+                }
+                functionCallFreeSet.addAll(set);
+            } else {
+                functionCallFreeSet.add(prev);
+            }
+        }
+        if (functionCallFreeSet.size() == 0) { // No node is found
+            return false;
+        } else if (isEmptySet) { // Set is not empty but at least a path has empty list. Should give up optimization.
+            return true;
+        }
+
+        EnumSet<MemoryAtomicityMode> modeSet = EnumSet.noneOf(MemoryAtomicityMode.class);
+        for (Node prev : functionCallFreeSet) {
+            modeSet.add(getMode(prev));
+        }
+
+        if (modeSet.contains(MemoryAtomicityMode.RELEASE) && !modeSet.contains(MemoryAtomicityMode.ACQUIRE)) {
+            // Weakest fence in modeSet is release. Check node n is also the release fence.
+            if (n instanceof Fence) {
```

Comment:
Is this saying that all paths to this node have at least a `RELEASE` fence immediately prior to this fence?  I'm not following the patterns being envisioned here (it might just be lack of imagination on my part)